src/lib/FreeC/LiftedIR/Syntax/Expr.hs

```diff
@@ -0,0 +1,116 @@
+-- | This module contains the definition of expressions for our lifted
+--   intermediate language.
+
+module FreeC.LiftedIR.Syntax.Expr where
+
+import           FreeC.IR.SrcSpan               ( SrcSpan )
+
+import           FreeC.LiftedIR.Effect          ( Effect )
+import           FreeC.LiftedIR.Syntax.Name
+import           FreeC.LiftedIR.Syntax.Type
+
+-- | An expression.
+data Expr
+  = -- | A constructor.
+    Con { exprSrcSpan :: SrcSpan
+        , exprConName :: ConName
+        , exprType    :: Type
+        }
+
+  | -- | A function or local variable.
+    Var { exprSrcSpan :: SrcSpan
+        , exprVarName :: VarName
+        , exprType    :: Type
+        }
+
+  | -- | Function or constructor application.
+    App { exprSrcSpan     :: SrcSpan
+        , exprAppFunc     :: Expr
+        , exprAppTypeArgs :: [Type]   -- ^ Visible type applications.
+        , exprEffects     :: [Effect] -- ^ Effect set.
+        , exprAppArgs     :: [Expr]   -- ^ Applied arguments.
+        , exprType        :: Type
+        }
+
+  | -- | @if@ expression.
+    If { exprSrcSpan :: SrcSpan
+       , ifExprCond  :: Expr
+       , ifExprThen  :: Expr
+       , ifExprElse  :: Expr
+       , exprType    :: Type
+       }
+
+  | -- | @case@ expression.
+    Case { exprSrcSpan       :: SrcSpan
+         , caseExprScrutinee :: Expr
+         , caseExprAlts      :: [Alt]
+         , exprType          :: Type
+         }
+
+  | -- | Error term @undefined@.
+    Undefined { exprSrcSpan :: SrcSpan
+              , exprType    :: Type
+              }
+
+  | -- | Error term @error "<message>"@.
+    ErrorExpr { exprSrcSpan  :: SrcSpan
+              , errorExprMsg :: String
+              , exprType     :: Type
+              }
+
+  | -- | An integer literal.
+    IntLiteral { exprSrcSpan     :: SrcSpan
+               , intLiteralValue :: Integer
+               , exprType        :: Type
+               }
+
+  | -- | A lambda abstraction.
+    Lambda { exprSrcSpan    :: SrcSpan
+           , lambdaExprArgs :: [VarPat]
+           , lambdaEprRhs   :: Expr
+           , exprType       :: Type
+           }
+
```

Comment:
I just remembered that `pure` is not only needed in the declaration of smart constructors but also around integer literals, lambda expression and (most importantly) strict arguments such as decreasing arguments that are used in positions where lifted values are expected. Thus, we still need a syntactic construct for `pure` :sweat_smile: 

```haskell
  | -- | The `pure` constructor of the free monad.
    Pure { exprSrcSpan  :: SrcSpan
         , exprPureArg  :: Expr -- ^ The value that is lifted into the @Free@ monad.
         , exprType     :: Type
         }
```